{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "98cc3a6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c8c35901",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "#import mitosheet\n",
    "#mitosheet.sheet()\n",
    "import scipy.stats as stats\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4f8a27e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_disease = pd.read_csv(\"all_variant_disease_associations.tsv.gz\", compression='gzip', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "345c5a52",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14155,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['1-3 toe syndactyly', '1-5 finger syndactyly', '11 pairs of ribs',\n",
       "       ..., 'von Willebrand disease type 2M',\n",
       "       \"von Willebrand's factor (lab test)\", 'winter depression'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(np.shape(np.unique(variant_disease['diseaseName'])))\n",
    "np.unique(variant_disease['diseaseName'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4c5957c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['Abnormality of the ovary',\n",
       "       'Adult Type Ovarian Granulosa Cell Tumor',\n",
       "       'Carcinoma, Ovarian Epithelial',\n",
       "       'Clear cell adenocarcinoma of ovary',\n",
       "       'Endometrioid carcinoma ovary', 'Epithelial ovarian cancer',\n",
       "       'Hereditary Breast and Ovarian Cancer Syndrome',\n",
       "       'Malignant Ovarian Germ Cell Neoplasm',\n",
       "       'Malignant neoplasm of ovary', 'Ovarian Carcinoma',\n",
       "       'Ovarian Granulosa Cell Tumor', 'Ovarian Mucinous Adenocarcinoma',\n",
       "       'Ovarian Serous Adenocarcinoma',\n",
       "       'Ovarian Serous Surface Papillary Adenocarcinoma',\n",
       "       'Ovarian adenocarcinoma', 'Ovarian germ cell tumour mixed',\n",
       "       'Ovarian mucinous tumor',\n",
       "       'Primary low grade serous adenocarcinoma of ovary',\n",
       "       'Recurrent ovarian cancer',\n",
       "       'Secondary malignant neoplasm of ovary',\n",
       "       'Serous cystadenocarcinoma ovary',\n",
       "       'Small cell carcinoma of the ovary, hypercalcemic type',\n",
       "       'Stromal tumor of ovary', 'ovarian neoplasm'], dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "OV = list()\n",
    "for i in range(len(variant_disease['diseaseName'])):\n",
    "    if 'Ovarian' in variant_disease['diseaseName'][i] and 'Failure' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Ovarian Cysts' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Ovarian Diseases' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Ovarian Dysgenesis 2' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Ovarian Hyperstimulation Syndrome' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Ovarian Reserve' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Ovarian dysgenesis' not in variant_disease['diseaseName'][i]\\\n",
    "    and 'Ovarian failure' not in variant_disease['diseaseName'][i]:\n",
    "        OV.append(i)\n",
    "    if 'ovarian' in variant_disease['diseaseName'][i] and 'Diminished' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Idiopathic premature ovarian failure' not in variant_disease['diseaseName'][i]:\n",
    "        OV.append(i)\n",
    "    if 'Ovary' in variant_disease['diseaseName'][i] and 'Gonadotropin-Resistant Ovary Syndrome' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Polycystic Ovary Syndrome' not in variant_disease['diseaseName'][i]:\n",
    "        OV.append(i)\n",
    "    if 'ovary' in variant_disease['diseaseName'][i] and 'Chocolate cyst of ovary' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Congenital absence of ovary' not in variant_disease['diseaseName'][i]\\\n",
    "    and 'Disorder of endocrine ovary' not in variant_disease['diseaseName'][i]\\\n",
    "    and 'Endometriosis of ovary' not in variant_disease['diseaseName'][i]:\n",
    "        OV.append(i)\n",
    "\n",
    "print(len(np.unique(variant_disease['diseaseName'][OV])))\n",
    "np.unique(variant_disease['diseaseName'][OV])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7fa6b142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snpId</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>DSI</th>\n",
       "      <th>DPI</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>diseaseName</th>\n",
       "      <th>diseaseType</th>\n",
       "      <th>diseaseClass</th>\n",
       "      <th>diseaseSemanticType</th>\n",
       "      <th>score</th>\n",
       "      <th>EI</th>\n",
       "      <th>YearInitial</th>\n",
       "      <th>YearFinal</th>\n",
       "      <th>NofPmids</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>rs10017134</td>\n",
       "      <td>4</td>\n",
       "      <td>69591303</td>\n",
       "      <td>0.925</td>\n",
       "      <td>0.12</td>\n",
       "      <td>C0677886</td>\n",
       "      <td>Epithelial ovarian cancer</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C13;C19</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>rs10017134</td>\n",
       "      <td>4</td>\n",
       "      <td>69591303</td>\n",
       "      <td>0.925</td>\n",
       "      <td>0.12</td>\n",
       "      <td>C4721610</td>\n",
       "      <td>Carcinoma, Ovarian Epithelial</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C13;C19</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>rs1002076</td>\n",
       "      <td>1</td>\n",
       "      <td>10378834</td>\n",
       "      <td>0.925</td>\n",
       "      <td>0.12</td>\n",
       "      <td>C0677886</td>\n",
       "      <td>Epithelial ovarian cancer</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C13;C19</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>rs1002076</td>\n",
       "      <td>1</td>\n",
       "      <td>10378834</td>\n",
       "      <td>0.925</td>\n",
       "      <td>0.12</td>\n",
       "      <td>C4721610</td>\n",
       "      <td>Carcinoma, Ovarian Epithelial</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C13;C19</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105</th>\n",
       "      <td>rs10069690</td>\n",
       "      <td>5</td>\n",
       "      <td>1279675</td>\n",
       "      <td>0.595</td>\n",
       "      <td>0.56</td>\n",
       "      <td>C0677886</td>\n",
       "      <td>Epithelial ovarian cancer</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C13;C19</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>1</td>\n",
       "      <td>GWASCAT;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367257</th>\n",
       "      <td>rs985033810</td>\n",
       "      <td>17</td>\n",
       "      <td>7674232</td>\n",
       "      <td>0.724</td>\n",
       "      <td>0.28</td>\n",
       "      <td>C0919267</td>\n",
       "      <td>ovarian neoplasm</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C13;C19</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>CLINVAR;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367529</th>\n",
       "      <td>rs9870207</td>\n",
       "      <td>3</td>\n",
       "      <td>190807727</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.04</td>\n",
       "      <td>C1335177</td>\n",
       "      <td>Ovarian Serous Adenocarcinoma</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>1</td>\n",
       "      <td>GWASCAT;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367751</th>\n",
       "      <td>rs9886651</td>\n",
       "      <td>8</td>\n",
       "      <td>127805637</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.12</td>\n",
       "      <td>C0919267</td>\n",
       "      <td>ovarian neoplasm</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C13;C19</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>1</td>\n",
       "      <td>GWASCAT;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368134</th>\n",
       "      <td>rs9909104</td>\n",
       "      <td>17</td>\n",
       "      <td>18344707</td>\n",
       "      <td>0.882</td>\n",
       "      <td>0.20</td>\n",
       "      <td>C0677886</td>\n",
       "      <td>Epithelial ovarian cancer</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C13;C19</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368135</th>\n",
       "      <td>rs9909104</td>\n",
       "      <td>17</td>\n",
       "      <td>18344707</td>\n",
       "      <td>0.882</td>\n",
       "      <td>0.20</td>\n",
       "      <td>C4721610</td>\n",
       "      <td>Carcinoma, Ovarian Epithelial</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C13;C19</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3908 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              snpId chromosome   position    DSI   DPI diseaseId  \\\n",
       "258      rs10017134          4   69591303  0.925  0.12  C0677886   \n",
       "259      rs10017134          4   69591303  0.925  0.12  C4721610   \n",
       "312       rs1002076          1   10378834  0.925  0.12  C0677886   \n",
       "313       rs1002076          1   10378834  0.925  0.12  C4721610   \n",
       "1105     rs10069690          5    1279675  0.595  0.56  C0677886   \n",
       "...             ...        ...        ...    ...   ...       ...   \n",
       "367257  rs985033810         17    7674232  0.724  0.28  C0919267   \n",
       "367529    rs9870207          3  190807727  1.000  0.04  C1335177   \n",
       "367751    rs9886651          8  127805637  1.000  0.12  C0919267   \n",
       "368134    rs9909104         17   18344707  0.882  0.20  C0677886   \n",
       "368135    rs9909104         17   18344707  0.882  0.20  C4721610   \n",
       "\n",
       "                          diseaseName diseaseType diseaseClass  \\\n",
       "258         Epithelial ovarian cancer     disease  C04;C13;C19   \n",
       "259     Carcinoma, Ovarian Epithelial     disease  C04;C13;C19   \n",
       "312         Epithelial ovarian cancer     disease  C04;C13;C19   \n",
       "313     Carcinoma, Ovarian Epithelial     disease  C04;C13;C19   \n",
       "1105        Epithelial ovarian cancer     disease  C04;C13;C19   \n",
       "...                               ...         ...          ...   \n",
       "367257               ovarian neoplasm     disease  C04;C13;C19   \n",
       "367529  Ovarian Serous Adenocarcinoma     disease          C04   \n",
       "367751               ovarian neoplasm     disease  C04;C13;C19   \n",
       "368134      Epithelial ovarian cancer     disease  C04;C13;C19   \n",
       "368135  Carcinoma, Ovarian Epithelial     disease  C04;C13;C19   \n",
       "\n",
       "       diseaseSemanticType  score   EI  YearInitial  YearFinal  NofPmids  \\\n",
       "258     Neoplastic Process   0.01  1.0       2019.0     2019.0         1   \n",
       "259     Neoplastic Process   0.01  1.0       2019.0     2019.0         1   \n",
       "312     Neoplastic Process   0.01  0.0       2015.0     2015.0         1   \n",
       "313     Neoplastic Process   0.01  0.0       2015.0     2015.0         1   \n",
       "1105    Neoplastic Process   0.70  1.0       2015.0     2015.0         1   \n",
       "...                    ...    ...  ...          ...        ...       ...   \n",
       "367257  Neoplastic Process   0.70  NaN          NaN        NaN         0   \n",
       "367529  Neoplastic Process   0.70  1.0       2017.0     2017.0         1   \n",
       "367751  Neoplastic Process   0.70  1.0       2017.0     2017.0         1   \n",
       "368134  Neoplastic Process   0.01  1.0       2008.0     2008.0         1   \n",
       "368135  Neoplastic Process   0.01  1.0       2008.0     2008.0         1   \n",
       "\n",
       "                  source  \n",
       "258               BEFREE  \n",
       "259               BEFREE  \n",
       "312               BEFREE  \n",
       "313               BEFREE  \n",
       "1105    GWASCAT;INFERRED  \n",
       "...                  ...  \n",
       "367257  CLINVAR;INFERRED  \n",
       "367529  GWASCAT;INFERRED  \n",
       "367751  GWASCAT;INFERRED  \n",
       "368134            BEFREE  \n",
       "368135            BEFREE  \n",
       "\n",
       "[3908 rows x 16 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#These are information about the diseases from the database\n",
    "variant_disease.loc[OV]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3967e736",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "376"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Map the database to the geneSymbol names\n",
    "snpId = variant_disease.loc[OV]['snpId']\n",
    "variant_gene = pd.read_csv(\"variant_to_gene_mappings.tsv.gz\", compression='gzip', sep = '\\t')\n",
    "chosen = list()\n",
    "for i in range(len(variant_gene['snpId'])):\n",
    "    if variant_gene['snpId'][i] in np.array(snpId):\n",
    "        chosen.append(i)\n",
    "variant_gene.loc[chosen]\n",
    "\n",
    "# Here is the list of GWAS genes that we are interested in\n",
    "GWAS_OV = np.unique(variant_gene.loc[chosen]['geneSymbol'])\n",
    "len(GWAS_OV)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "14cbfbe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''GWAS enrichment'''\n",
    "'''DEGs are selected all DSs'''\n",
    "    \n",
    "#DEGs of each celltype \n",
    "cancer_list = ['Lung','Liver','Breast','Colon','Ovary']\n",
    "cancer = \"Ovary\" # change here for different cancers\n",
    "\n",
    "with open('/GWAS_enrich_percelltype_'+cancer+'.txt','w') as f:\n",
    "\n",
    "    DEGs = pd.read_csv('/ligand_target_merged_withFC.csv',sep = ',')\n",
    "    DEGs = DEGs.loc[DEGs['cancer']==cancer]\n",
    "    DEGs\n",
    "\n",
    "    celltype_DEG = np.unique(DEGs['receiver'])\n",
    "    print(\"celltypes_DEGs: \", celltype_DEG, file =f)\n",
    "\n",
    "    ''' background genes'''\n",
    "    bg_file_path = (\"/background_genes_\"+cancer+\".csv\")\n",
    "    all_gene = np.arange(0)\n",
    "    all = pd.read_csv(bg_file_path,sep = ',') \n",
    "    all = all.drop('Unnamed: 0',axis = 1)\n",
    "\n",
    "    celltype_bg = all.columns\n",
    "    print(\"celltypes_bgs: \", celltype_bg, file =f)\n",
    "    DEGs\n",
    "\n",
    "    '''loop all celltype'''\n",
    "    for c in range(len(celltype_DEG)):\n",
    "        print('\\n', file =f)\n",
    "        print(celltype_DEG[c], file =f)\n",
    "        iDEGs = DEGs.loc[DEGs['receiver']==celltype_DEG[c]]['target'].unique()\n",
    "        print(celltype_DEG[c],\"DEGs\",len(iDEGs), file =f)\n",
    "        ibackground_genes = all[celltype_DEG[c]].unique()\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'NAN']\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'nan']\n",
    "        print(celltype_DEG[c],\"background\",len(ibackground_genes), file =f)\n",
    "\n",
    "        GWAS = list()\n",
    "        for g in GWAS_OV:\n",
    "            for a in ibackground_genes:\n",
    "                a = str(a)\n",
    "                for i in range(len(a.split('///'))):\n",
    "                    if g == a.split('///')[i]:\n",
    "                        GWAS.append(a)                   \n",
    "        GWAS = np.unique(GWAS)\n",
    "        GWAS\n",
    "\n",
    "        nonDEGs = list(set(ibackground_genes) - set(iDEGs))                \n",
    "        table11 = len(np.intersect1d(GWAS, iDEGs))\n",
    "        table12 = len(iDEGs) - table11\n",
    "        table21 = len(np.intersect1d(GWAS, nonDEGs))\n",
    "        table22 = len(ibackground_genes) - (table21 + table12 + table11)\n",
    "        table = pd.DataFrame([[table11, table12], [table21, table22]])\n",
    "        table.index = ('DEGs', 'nonDEGs')\n",
    "        table.columns = ('GWAS', 'nonGWAS')\n",
    "\n",
    "        print('Ovarian Cancer - ',celltype_DEG[c], file =f )\n",
    "        print(table, file =f)\n",
    "        odds, pval = stats.fisher_exact(table)\n",
    "        print(\"Odds ratio: \", round(odds, 4), file =f)\n",
    "        print(\"p-value: \",round(pval, 6), file =f)\n",
    "\n",
    "        # if the odds are > 1 and pval <0.05 The list of DEGs is significantly enriched by the GWAS genes\n",
    "        # if the odds are < 1 and pval <0.05 The list of DEGs is not enriched by the GWAS genes\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "066b9194",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "celltypes_bgs:  Index(['B', 'CD4T', 'CD8T', 'DC', 'Endothelial', 'Epithelial', 'Fibroblast',\n",
      "       'Macrophage', 'Mast', 'Monocyte', 'NK', 'NKT', 'Pericyte', 'Plasma'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "    bg_file_path = (\"/DEGs_of_\"+cancer+\"/DEGs_merged_\"+cancer+\"_MAST.txt\")\n",
    "    all_gene = np.arange(0)\n",
    "    all = pd.read_csv(bg_file_path,sep = '\\t') \n",
    "    all\n",
    "    \n",
    "    celltype_bg = all.columns\n",
    "    print(\"celltypes_bgs: \", celltype_bg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e95ae681",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''GWAS enrichment'''\n",
    "'''DEGs are selected all DSs\n",
    "    background genes are all differentially expressed genes\n",
    "'''\n",
    "    \n",
    "#DEGs of each celltype \n",
    "cancer_list = ['Lung','Liver','Breast','Colon','Ovary']\n",
    "cancer = \"Ovary\" # change here for different cancers\n",
    "\n",
    "with open('/GWAS_enrich_percelltype_DEGasBG_'+cancer+'.txt','w') as f:\n",
    "\n",
    "    DEGs = pd.read_csv('/ligand_target_merged_withFC.csv',sep = ',')\n",
    "    DEGs = DEGs.loc[DEGs['cancer']==cancer]\n",
    "    DEGs\n",
    "\n",
    "    celltype_DEG = np.unique(DEGs['receiver'])\n",
    "    print(\"celltypes_DEGs: \", celltype_DEG, file =f)\n",
    "\n",
    "    '''Differentially expressed genes in each cell type were background genes'''\n",
    "    bg_file_path = (\"/DEGs_of_\"+cancer+\"/DEGs_merged_\"+cancer+\"_MAST.txt\")\n",
    "    all_gene = np.arange(0)\n",
    "    all = pd.read_csv(bg_file_path,sep = '\\t') \n",
    "    all\n",
    "    \n",
    "    celltype_bg = all.columns\n",
    "    print(\"celltypes_bgs: \", celltype_bg, file =f)\n",
    "    DEGs\n",
    "\n",
    "    '''for Ovary only - loop all celltype'''\n",
    "    for c in range(len(celltype_DEG)):\n",
    "        print('\\n', file =f)\n",
    "        print(celltype_DEG[c], file =f)\n",
    "        iDEGs = DEGs.loc[DEGs['receiver']==celltype_DEG[c]]['target'].unique()\n",
    "        print(celltype_DEG[c],\"DEGs\",len(iDEGs), file =f)\n",
    "        ibackground_genes = all[celltype_DEG[c]].unique()\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'NAN']\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'nan']\n",
    "        print(celltype_DEG[c],\"background\",len(ibackground_genes), file =f)\n",
    "\n",
    "        GWAS = list()\n",
    "        for g in GWAS_OV:\n",
    "            for a in ibackground_genes:\n",
    "                a = str(a)\n",
    "                for i in range(len(a.split('///'))):\n",
    "                    if g == a.split('///')[i]:\n",
    "                        GWAS.append(a)                   \n",
    "        GWAS = np.unique(GWAS)\n",
    "        GWAS\n",
    "\n",
    "        nonDEGs = list(set(ibackground_genes) - set(iDEGs))                \n",
    "        table11 = len(np.intersect1d(GWAS, iDEGs))\n",
    "        table12 = len(iDEGs) - table11\n",
    "        table21 = len(np.intersect1d(GWAS, nonDEGs))\n",
    "        table22 = len(ibackground_genes) - (table21 + table12 + table11)\n",
    "        table = pd.DataFrame([[table11, table12], [table21, table22]])\n",
    "        table.index = ('DEGs', 'nonDEGs')\n",
    "        table.columns = ('GWAS', 'nonGWAS')\n",
    "\n",
    "        print('Ovarian Cancer - ',celltype_DEG[c], file =f )\n",
    "        print(table, file =f)\n",
    "        odds, pval = stats.fisher_exact(table)\n",
    "        print(\"Odds ratio: \", round(odds, 4), file =f)\n",
    "        print(\"p-value: \",round(pval, 6), file =f)\n",
    "\n",
    "        # if the odds are > 1 and pval <0.05 The list of DEGs is significantly enriched by the GWAS genes\n",
    "        # if the odds are < 1 and pval <0.05 The list of DEGs is not enriched by the GWAS genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8a0a4ae4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ANXA1' 'APOD' 'CFD' 'CNN2' 'CRIP1' 'CXCR4' 'CYR61' 'DCN' 'DUSP1'\n",
      " 'EEF1A1' 'ETS1' 'FABP4' 'FOS' 'GAPDH' 'GPR183' 'GZMB' 'HBA1' 'HBA2'\n",
      " 'IGHA1' 'JUNB' 'KLF6' 'MCL1' 'MGP' 'NEAT1' 'PIK3R1' 'PTGER4' 'RPS29'\n",
      " 'SPP1' 'SRGN' 'TNFAIP3' 'TSC22D3']\n",
      "['CXCR4' 'DUSP1' 'EEF1A1' 'ETS1' 'FOS' 'GAPDH' 'GZMB' 'HBA1' 'HBA2' 'JUNB'\n",
      " 'KLF6' 'MCL1' 'NEAT1' 'PIK3R1' 'SPP1' 'TNFAIP3' 'TSC22D3']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "    DEGs = pd.read_csv('/ligand_target_merged_withFC.csv',sep = ',')\n",
    "    DEGs = DEGs.loc[DEGs['cancer']==cancer]\n",
    "    DEGs = DEGs.loc[DEGs['receiver']=='CD4T']['target']\n",
    "    DEGs = np.unique(DEGs)\n",
    "    print(DEGs)\n",
    "    \n",
    "    DEGs = pd.read_csv('/ligand_target_merged_DS_of_scUR.csv',sep = ',')\n",
    "    DEGs = DEGs.loc[DEGs['cancer']==cancer]\n",
    "    DEGs = DEGs.loc[DEGs['receiver']=='CD4T']['target']\n",
    "    DEGs = np.unique(DEGs)\n",
    "    print(DEGs)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "829fa577",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''GWAS enrichment'''\n",
    "'''DEGs are selected all DSs of the top scURs\n",
    "    background genes are all differentially expressed genes\n",
    "'''\n",
    "    \n",
    "#DEGs of each celltype \n",
    "cancer_list = ['Lung','Liver','Breast','Colon','Ovary']\n",
    "cancer = \"Ovary\" # change here for different cancers\n",
    "\n",
    "with open('/GWAS_enrich_percelltype_DS4topscUR_DEGasBG_'+cancer+'.txt','w') as f:\n",
    "\n",
    "    DEGs = pd.read_csv('/ligand_target_merged_DS_of_scUR.csv',sep = ',')\n",
    "    top_scURs = [\"COL1A1\",\"FN1\",\"SPP1\",\"COL4A1\",\"COL18A1\", \"PLAU\" , \"CLEC11A\" ,\"MDK\"]\n",
    "    DEGs = DEGs.loc[np.isin(DEGs['ligand'], top_scURs)]\n",
    "    DEGs = DEGs.loc[DEGs['cancer']==cancer]\n",
    "\n",
    "    celltype_DEG = np.unique(DEGs['receiver'])\n",
    "    print(\"celltypes_DEGs: \", celltype_DEG, file =f)\n",
    "\n",
    "    '''Differentially expressed genes in each cell type were background genes'''\n",
    "    bg_file_path = (\"/DEGs_of_\"+cancer+\"/DEGs_merged_\"+cancer+\"_MAST.txt\")\n",
    "    all_gene = np.arange(0)\n",
    "    all = pd.read_csv(bg_file_path,sep = '\\t') \n",
    "    all\n",
    "    \n",
    "    celltype_bg = all.columns\n",
    "    print(\"celltypes_bgs: \", celltype_bg, file =f)\n",
    "    DEGs\n",
    "\n",
    "    '''for Ovary only - loop all celltype'''\n",
    "    for c in range(len(celltype_DEG)):\n",
    "        print('\\n', file =f)\n",
    "        print(celltype_DEG[c], file =f)\n",
    "        iDEGs = DEGs.loc[DEGs['receiver']==celltype_DEG[c]]['target'].unique()\n",
    "        print(celltype_DEG[c],\"DEGs\",len(iDEGs), file =f)\n",
    "        ibackground_genes = all[celltype_DEG[c]].unique()\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'NAN']\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'nan']\n",
    "        print(celltype_DEG[c],\"background\",len(ibackground_genes), file =f)\n",
    "\n",
    "        GWAS = list()\n",
    "        for g in GWAS_OV:\n",
    "            for a in ibackground_genes:\n",
    "                a = str(a)\n",
    "                for i in range(len(a.split('///'))):\n",
    "                    if g == a.split('///')[i]:\n",
    "                        GWAS.append(a)                   \n",
    "        GWAS = np.unique(GWAS)\n",
    "        GWAS\n",
    "\n",
    "        nonDEGs = list(set(ibackground_genes) - set(iDEGs))                \n",
    "        table11 = len(np.intersect1d(GWAS, iDEGs))\n",
    "        table12 = len(iDEGs) - table11\n",
    "        table21 = len(np.intersect1d(GWAS, nonDEGs))\n",
    "        table22 = len(ibackground_genes) - (table21 + table12 + table11)\n",
    "        table = pd.DataFrame([[table11, table12], [table21, table22]])\n",
    "        table.index = ('DEGs', 'nonDEGs')\n",
    "        table.columns = ('GWAS', 'nonGWAS')\n",
    "\n",
    "        print('Ovarian Cancer - ',celltype_DEG[c], file =f )\n",
    "        print(table, file =f)\n",
    "        odds, pval = stats.fisher_exact(table)\n",
    "        print(\"Odds ratio: \", round(odds, 4), file =f)\n",
    "        print(\"p-value: \",round(pval, 6), file =f)\n",
    "\n",
    "        # if the odds are > 1 and pval <0.05 The list of DEGs is significantly enriched by the GWAS genes\n",
    "        # if the odds are < 1 and pval <0.05 The list of DEGs is not enriched by the GWAS genes\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "895042a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['COL18A1', 'KRT8', 'PLAU'], dtype=object)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list GWAS gene that in our gene-of-interests\n",
    "np.intersect1d(GWAS, DEGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd2ee623",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
