{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d1d2306d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "#import mitosheet\n",
    "#mitosheet.sheet()\n",
    "import scipy.stats as stats\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "46ba44b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_disease = pd.read_csv(\"all_variant_disease_associations.tsv.gz\", compression='gzip', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d7eea78c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14155,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['1-3 toe syndactyly', '1-5 finger syndactyly', '11 pairs of ribs',\n",
       "       ..., 'von Willebrand disease type 2M',\n",
       "       \"von Willebrand's factor (lab test)\", 'winter depression'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(np.shape(np.unique(variant_disease['diseaseName'])))\n",
    "np.unique(variant_disease['diseaseName'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5832a4b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['Adult Hepatocellular Carcinoma', 'Adult Liver Carcinoma',\n",
       "       'Childhood Hepatocellular Carcinoma',\n",
       "       'Clear Cell Hepatocellular Carcinoma', 'End Stage Liver Disease',\n",
       "       'Extrahepatic Cholangiocarcinoma',\n",
       "       'Hepatitis B Virus-Related Hepatocellular Carcinoma',\n",
       "       'Hepatitis C Virus-Related Hepatocellular Carcinoma',\n",
       "       'Hepatitis Virus-Related Hepatocellular Carcinoma',\n",
       "       'Hepatobiliary cancer', 'Hepatoblastoma', 'Hepatocarcinogenesis',\n",
       "       'Hepatocellular Adenoma', 'Intrahepatic Cholangiocarcinoma',\n",
       "       'Liver and Intrahepatic Biliary Tract Carcinoma',\n",
       "       'Liver carcinoma', 'Liver neoplasms',\n",
       "       'Malignant neoplasm of liver', 'Primary Malignant Liver Neoplasm',\n",
       "       'Primary cholangiocarcinoma of intrahepatic biliary tract',\n",
       "       'Sarcoma of liver', 'Secondary malignant neoplasm of liver'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Liver = list()\n",
    "for i in range(len(variant_disease['diseaseName'])):\n",
    "    if 'Liver' in variant_disease['diseaseName'][i] \\\n",
    "    and 'Liver Failure' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Fatty' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Cirrhosis' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Alcoholic Liver Diseases' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Chemical and Drug Induced Liver Injury' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Fibrosis, Liver' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'cyst' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'disease' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Abscess' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Drug-Induced'  not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Glycogen' not in variant_disease['diseaseName'][i] \\\n",
    "    :\n",
    "        Liver.append(i)\n",
    "    if 'liver' in variant_disease['diseaseName'][i] \\\n",
    "    and 'delivery' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'pregnancy' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Chronic liver disease' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Congenital absence of liver' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'cirrhosis' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Abnormality' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Oliver' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Chronic' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'chronic' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Angiosarcoma of liver' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Autoimmune liver disease' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'function' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Drug-Induced' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'enzyme' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'cyst' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'alcohol' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Hyperbiliverdinemia' not in variant_disease['diseaseName'][i] \\\n",
    "    :\n",
    "        Liver.append(i)\n",
    "    if 'Hepatocellular Carcinoma' in variant_disease['diseaseName'][i] :\n",
    "        Liver.append(i)  \n",
    "    if 'Hepa' in variant_disease['diseaseName'][i] \\\n",
    "    and 'Heparin' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Chronic' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'chronic' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Autoimmune Hepatitis with Centrilobular Necrosis' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'siderosis' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Alcoholic' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Cryohydrocytosis' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Echinococcosis' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Hepatic' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Hepatitis' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Measurement' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Hepatoerythropoietic Porphyria' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Hepatolenticular Degeneration' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Hepatomegaly' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Hepatoportal sclerosis' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Hepatopulmonary Syndrome' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Hepatosplenomegaly' not in variant_disease['diseaseName'][i] \\\n",
    "    :\n",
    "        Liver.append(i)\n",
    "    if 'hepa' in variant_disease['diseaseName'][i] \\\n",
    "    and 'Heparin' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Cholestasis' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Abnormality' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Acute' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Chronic' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'chronic' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Autoimmune hepatitis' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Benign recurrent intrahepatic cholestasis' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Congenital atresia of extrahepatic bile duct' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'hepatitis' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Elevated' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'cyst' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'disorder' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Reduced number of intrahepatic bile ducts'  not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Subacute hepatic necrosis' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'Trichohepatoenteric Syndrome' not in variant_disease['diseaseName'][i] \\\n",
    "    and 'response to heparin' not in variant_disease['diseaseName'][i] \\\n",
    "    :\n",
    "        Liver.append(i)\n",
    "\n",
    "print(len(np.unique(variant_disease['diseaseName'][Liver])))\n",
    "np.unique(variant_disease['diseaseName'][Liver])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "662a62c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#These are information about the diseases from the database\n",
    "# variant_disease.loc[Colon]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "157f3d0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "658\n"
     ]
    }
   ],
   "source": [
    "#Map the database to the geneSymbol names\n",
    "snpId = variant_disease.loc[Liver]['snpId']\n",
    "variant_gene = pd.read_csv(\"variant_to_gene_mappings.tsv.gz\", compression='gzip', sep = '\\t')\n",
    "chosen = list()\n",
    "for i in range(len(variant_gene['snpId'])):\n",
    "    if variant_gene['snpId'][i] in np.array(snpId):\n",
    "        chosen.append(i)\n",
    "variant_gene.loc[chosen]\n",
    "\n",
    "# Here is the list of GWAS genes that we are interested in\n",
    "GWAS_Liver = np.unique(variant_gene.loc[chosen]['geneSymbol'])\n",
    "print(len(GWAS_Liver))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d50c80c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''GWAS enrichment'''\n",
    "'''DEGs are selected DSs'''\n",
    "    \n",
    "#DEGs of each celltype \n",
    "cancer_list = ['Lung','Liver','Breast','Colon','Ovary']\n",
    "cancer = \"Liver\" # change here for different cancers\n",
    "\n",
    "with open('/GWAS_enrich_percelltype_'+cancer+'.txt','w') as f:\n",
    "\n",
    "    DEGs = pd.read_csv('/ligand_target_merged_withFC.csv',sep = ',')\n",
    "    DEGs = DEGs.loc[DEGs['cancer']==cancer]\n",
    "    DEGs\n",
    "\n",
    "    celltype_DEG = np.unique(DEGs['receiver'])\n",
    "    print(\"celltypes_DEGs: \", celltype_DEG, file =f)\n",
    "\n",
    "    '''all_genes means the combine of all background genes in all celltypes'''\n",
    "    bg_file_path = (\"/background_genes_\"+cancer+\".csv\")\n",
    "    all_gene = np.arange(0)\n",
    "    all = pd.read_csv(bg_file_path,sep = ',') \n",
    "    all = all.drop('Unnamed: 0',axis = 1)\n",
    "\n",
    "    celltype_bg = all.columns\n",
    "    print(\"celltypes_bgs: \", celltype_bg, file =f)\n",
    "    DEGs\n",
    "\n",
    "    '''for Lung only - loop all celltype'''\n",
    "    for c in range(len(celltype_DEG)):\n",
    "        print('\\n', file =f)\n",
    "        print(celltype_DEG[c], file =f)\n",
    "        iDEGs = DEGs.loc[DEGs['receiver']==celltype_DEG[c]]['target'].unique()\n",
    "        print(celltype_DEG[c],\"DEGs\",len(iDEGs), file =f)\n",
    "        ibackground_genes = all[celltype_DEG[c]].unique()\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'NAN']\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'nan']\n",
    "        print(celltype_DEG[c],\"background\",len(ibackground_genes), file =f)\n",
    "\n",
    "        GWAS = list()\n",
    "        for g in GWAS_Liver:\n",
    "            for a in ibackground_genes:\n",
    "                a = str(a)\n",
    "                for i in range(len(a.split('///'))):\n",
    "                    if g == a.split('///')[i]:\n",
    "                        GWAS.append(a)                   \n",
    "        GWAS = np.unique(GWAS)\n",
    "        GWAS\n",
    "\n",
    "        nonDEGs = list(set(ibackground_genes) - set(iDEGs))                \n",
    "        table11 = len(np.intersect1d(GWAS, iDEGs))\n",
    "        table12 = len(iDEGs) - table11\n",
    "        table21 = len(np.intersect1d(GWAS, nonDEGs))\n",
    "        table22 = len(ibackground_genes) - (table21 + table12 + table11)\n",
    "        table = pd.DataFrame([[table11, table12], [table21, table22]])\n",
    "        table.index = ('DEGs', 'nonDEGs')\n",
    "        table.columns = ('GWAS', 'nonGWAS')\n",
    "\n",
    "        print('Lung Cancer - ',celltype_DEG[c], file =f )\n",
    "        print(table, file =f)\n",
    "        odds, pval = stats.fisher_exact(table)\n",
    "        print(\"Odds ratio: \", round(odds, 4), file =f)\n",
    "        print(\"p-value: \",round(pval, 6), file =f)\n",
    "\n",
    "        # if the odds are > 1 and pval <0.05 The list of DEGs is significantly enriched by the GWAS genes\n",
    "        # if the odds are < 1 and pval <0.05 The list of DEGs is not enriched by the GWAS genes\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "66ac4a6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''GWAS enrichment'''\n",
    "'''DEGs are selected all DSs\n",
    "    background genes are all differentially expressed genes\n",
    "'''\n",
    "    \n",
    "#DEGs of each celltype \n",
    "cancer_list = ['Lung','Liver','Breast','Colon','Ovary']\n",
    "cancer = \"Liver\" # change here for different cancers\n",
    "\n",
    "with open('/GWAS_enrich_percelltype_DEGasBG_'+cancer+'.txt','w') as f:\n",
    "\n",
    "    DEGs = pd.read_csv('/ligand_target_merged_withFC.csv',sep = ',')\n",
    "    DEGs = DEGs.loc[DEGs['cancer']==cancer]\n",
    "    DEGs\n",
    "\n",
    "    celltype_DEG = np.unique(DEGs['receiver'])\n",
    "    print(\"celltypes_DEGs: \", celltype_DEG, file =f)\n",
    "\n",
    "    '''Differentially expressed genes in each cell type were background genes'''\n",
    "    bg_file_path = (\"/DEGs_of_\"+cancer+\"/DEGs_merged_\"+cancer+\"_MAST.txt\")\n",
    "    all_gene = np.arange(0)\n",
    "    all = pd.read_csv(bg_file_path,sep = '\\t') \n",
    "    all\n",
    "    \n",
    "    celltype_bg = all.columns\n",
    "    print(\"celltypes_bgs: \", celltype_bg, file =f)\n",
    "    DEGs\n",
    "\n",
    "    '''loop all celltype'''\n",
    "    for c in range(len(celltype_DEG)):\n",
    "        print('\\n', file =f)\n",
    "        print(celltype_DEG[c], file =f)\n",
    "        iDEGs = DEGs.loc[DEGs['receiver']==celltype_DEG[c]]['target'].unique()\n",
    "        print(celltype_DEG[c],\"DEGs\",len(iDEGs), file =f)\n",
    "        ibackground_genes = all[celltype_DEG[c]].unique()\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'NAN']\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'nan']\n",
    "        print(celltype_DEG[c],\"background\",len(ibackground_genes), file =f)\n",
    "\n",
    "        GWAS = list()\n",
    "        for g in GWAS_Liver:\n",
    "            for a in ibackground_genes:\n",
    "                a = str(a)\n",
    "                for i in range(len(a.split('///'))):\n",
    "                    if g == a.split('///')[i]:\n",
    "                        GWAS.append(a)                   \n",
    "        GWAS = np.unique(GWAS)\n",
    "        GWAS\n",
    "\n",
    "        nonDEGs = list(set(ibackground_genes) - set(iDEGs))                \n",
    "        table11 = len(np.intersect1d(GWAS, iDEGs))\n",
    "        table12 = len(iDEGs) - table11\n",
    "        table21 = len(np.intersect1d(GWAS, nonDEGs))\n",
    "        table22 = len(ibackground_genes) - (table21 + table12 + table11)\n",
    "        table = pd.DataFrame([[table11, table12], [table21, table22]])\n",
    "        table.index = ('DEGs', 'nonDEGs')\n",
    "        table.columns = ('GWAS', 'nonGWAS')\n",
    "\n",
    "        print('Liver Cancer - ',celltype_DEG[c], file =f )\n",
    "        print(table, file =f)\n",
    "        odds, pval = stats.fisher_exact(table)\n",
    "        print(\"Odds ratio: \", round(odds, 4), file =f)\n",
    "        print(\"p-value: \",round(pval, 6), file =f)\n",
    "\n",
    "        # if the odds are > 1 and pval <0.05 The list of DEGs is significantly enriched by the GWAS genes\n",
    "        # if the odds are < 1 and pval <0.05 The list of DEGs is not enriched by the GWAS genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f884607c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['BAX', 'CCND1', 'GADD45B', 'GNAS', 'KLF6', 'KRT8', 'MMP2', 'MMP9',\n",
       "       'MUC1', 'NR3C1', 'PLAU', 'SPP1', 'TGFB1', 'TIMP1'], dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list GWAS gene that in our gene-of-interests\n",
    "np.intersect1d(GWAS, DEGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a83adffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''GWAS enrichment'''\n",
    "'''DEGs are selected all DSs of the top scURs\n",
    "    background genes are all differentially expressed genes\n",
    "'''\n",
    "    \n",
    "#DEGs of each celltype \n",
    "cancer_list = ['Lung','Liver','Breast','Colon','Ovary']\n",
    "cancer = \"Liver\" # change here for different cancers\n",
    "\n",
    "with open('/GWAS_enrich_percelltype_DS4topscUR_DEGasBG_'+cancer+'.txt','w') as f:\n",
    "\n",
    "    DEGs = pd.read_csv('/ligand_target_merged_DS_of_scUR.csv',sep = ',')\n",
    "    top_scURs = [\"COL1A1\",\"FN1\",\"SPP1\",\"COL4A1\",\"COL18A1\", \"PLAU\" , \"CLEC11A\" ,\"MDK\"]\n",
    "    DEGs = DEGs.loc[np.isin(DEGs['ligand'], top_scURs)]\n",
    "    DEGs = DEGs.loc[DEGs['cancer']==cancer]\n",
    "\n",
    "\n",
    "    celltype_DEG = np.unique(DEGs['receiver'])\n",
    "    print(\"celltypes_DEGs: \", celltype_DEG, file =f)\n",
    "\n",
    "    '''Differentially expressed genes in each cell type were background genes'''\n",
    "    bg_file_path = (\"/DEGs_of_\"+cancer+\"/DEGs_merged_\"+cancer+\"_MAST.txt\")\n",
    "    all_gene = np.arange(0)\n",
    "    all = pd.read_csv(bg_file_path,sep = '\\t') \n",
    "    all\n",
    "    \n",
    "    celltype_bg = all.columns\n",
    "    print(\"celltypes_bgs: \", celltype_bg, file =f)\n",
    "    DEGs\n",
    "\n",
    "    '''loop all celltype'''\n",
    "    for c in range(len(celltype_DEG)):\n",
    "        print('\\n', file =f)\n",
    "        print(celltype_DEG[c], file =f)\n",
    "        iDEGs = DEGs.loc[DEGs['receiver']==celltype_DEG[c]]['target'].unique()\n",
    "        print(celltype_DEG[c],\"DEGs\",len(iDEGs), file =f)\n",
    "        ibackground_genes = all[celltype_DEG[c]].unique()\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'NAN']\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'nan']\n",
    "        print(celltype_DEG[c],\"background\",len(ibackground_genes), file =f)\n",
    "\n",
    "        GWAS = list()\n",
    "        for g in GWAS_Liver:\n",
    "            for a in ibackground_genes:\n",
    "                a = str(a)\n",
    "                for i in range(len(a.split('///'))):\n",
    "                    if g == a.split('///')[i]:\n",
    "                        GWAS.append(a)                   \n",
    "        GWAS = np.unique(GWAS)\n",
    "        GWAS\n",
    "\n",
    "        nonDEGs = list(set(ibackground_genes) - set(iDEGs))                \n",
    "        table11 = len(np.intersect1d(GWAS, iDEGs))\n",
    "        table12 = len(iDEGs) - table11\n",
    "        table21 = len(np.intersect1d(GWAS, nonDEGs))\n",
    "        table22 = len(ibackground_genes) - (table21 + table12 + table11)\n",
    "        table = pd.DataFrame([[table11, table12], [table21, table22]])\n",
    "        table.index = ('DEGs', 'nonDEGs')\n",
    "        table.columns = ('GWAS', 'nonGWAS')\n",
    "\n",
    "        print('Liver Cancer - ',celltype_DEG[c], file =f )\n",
    "        print(table, file =f)\n",
    "        odds, pval = stats.fisher_exact(table)\n",
    "        print(\"Odds ratio: \", round(odds, 4), file =f)\n",
    "        print(\"p-value: \",round(pval, 6), file =f)\n",
    "\n",
    "        # if the odds are > 1 and pval <0.05 The list of DEGs is significantly enriched by the GWAS genes\n",
    "        # if the odds are < 1 and pval <0.05 The list of DEGs is not enriched by the GWAS genes\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "463e4b18",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
