{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "#import mitosheet\n",
    "#mitosheet.sheet()\n",
    "import scipy.stats as stats\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_disease = pd.read_csv(\"all_variant_disease_associations.tsv.gz\", compression='gzip', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14155,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['1-3 toe syndactyly', '1-5 finger syndactyly', '11 pairs of ribs',\n",
       "       ..., 'von Willebrand disease type 2M',\n",
       "       \"von Willebrand's factor (lab test)\", 'winter depression'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(np.shape(np.unique(variant_disease['diseaseName'])))\n",
    "np.unique(variant_disease['diseaseName'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snpId</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>DSI</th>\n",
       "      <th>DPI</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>diseaseName</th>\n",
       "      <th>diseaseType</th>\n",
       "      <th>diseaseClass</th>\n",
       "      <th>diseaseSemanticType</th>\n",
       "      <th>score</th>\n",
       "      <th>EI</th>\n",
       "      <th>YearInitial</th>\n",
       "      <th>YearFinal</th>\n",
       "      <th>NofPmids</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1000005</td>\n",
       "      <td>21</td>\n",
       "      <td>33060745</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C0200638</td>\n",
       "      <td>Eosinophil count procedure</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Laboratory Procedure</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>1</td>\n",
       "      <td>GWASCAT;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs10000770</td>\n",
       "      <td>4</td>\n",
       "      <td>142693109</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.04</td>\n",
       "      <td>C0023467</td>\n",
       "      <td>Leukemia, Myelocytic, Acute</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>1</td>\n",
       "      <td>GWASCAT;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1000091588</td>\n",
       "      <td>11</td>\n",
       "      <td>68935374</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.12</td>\n",
       "      <td>C1858517</td>\n",
       "      <td>SPINAL MUSCULAR ATROPHY WITH RESPIRATORY DISTR...</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C08;C10</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>1</td>\n",
       "      <td>CLINVAR;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1000096</td>\n",
       "      <td>4</td>\n",
       "      <td>38691214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C1305855</td>\n",
       "      <td>Body mass index</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Clinical Attribute</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>1</td>\n",
       "      <td>GWASCAT;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs10001106</td>\n",
       "      <td>4</td>\n",
       "      <td>10125817</td>\n",
       "      <td>0.925</td>\n",
       "      <td>0.12</td>\n",
       "      <td>C0003868</td>\n",
       "      <td>Arthritis, Gouty</td>\n",
       "      <td>disease</td>\n",
       "      <td>C16;C18;C05</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>3</td>\n",
       "      <td>GWASDB;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369549</th>\n",
       "      <td>rs999947969</td>\n",
       "      <td>7</td>\n",
       "      <td>80671145</td>\n",
       "      <td>0.827</td>\n",
       "      <td>0.08</td>\n",
       "      <td>C0155626</td>\n",
       "      <td>Acute myocardial infarction</td>\n",
       "      <td>disease</td>\n",
       "      <td>C23;C14</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369550</th>\n",
       "      <td>rs999947969</td>\n",
       "      <td>7</td>\n",
       "      <td>80671145</td>\n",
       "      <td>0.827</td>\n",
       "      <td>0.08</td>\n",
       "      <td>C1956346</td>\n",
       "      <td>Coronary Artery Disease</td>\n",
       "      <td>disease</td>\n",
       "      <td>C14</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369551</th>\n",
       "      <td>rs9999653</td>\n",
       "      <td>4</td>\n",
       "      <td>55632082</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.08</td>\n",
       "      <td>C2362324</td>\n",
       "      <td>Pediatric Obesity</td>\n",
       "      <td>disease</td>\n",
       "      <td>C23;C18</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369552</th>\n",
       "      <td>rs9999820</td>\n",
       "      <td>4</td>\n",
       "      <td>155197173</td>\n",
       "      <td>0.925</td>\n",
       "      <td>0.08</td>\n",
       "      <td>C0025322</td>\n",
       "      <td>Premature Menopause</td>\n",
       "      <td>disease</td>\n",
       "      <td>C13</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369553</th>\n",
       "      <td>rs9999820</td>\n",
       "      <td>4</td>\n",
       "      <td>155197173</td>\n",
       "      <td>0.925</td>\n",
       "      <td>0.08</td>\n",
       "      <td>C0085215</td>\n",
       "      <td>Ovarian Failure, Premature</td>\n",
       "      <td>disease</td>\n",
       "      <td>C13;C19</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>1</td>\n",
       "      <td>GWASCAT;INFERRED</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>369554 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               snpId chromosome   position    DSI   DPI diseaseId  \\\n",
       "0          rs1000005         21   33060745    NaN   NaN  C0200638   \n",
       "1         rs10000770          4  142693109  1.000  0.04  C0023467   \n",
       "2       rs1000091588         11   68935374  1.000  0.12  C1858517   \n",
       "3          rs1000096          4   38691214    NaN   NaN  C1305855   \n",
       "4         rs10001106          4   10125817  0.925  0.12  C0003868   \n",
       "...              ...        ...        ...    ...   ...       ...   \n",
       "369549   rs999947969          7   80671145  0.827  0.08  C0155626   \n",
       "369550   rs999947969          7   80671145  0.827  0.08  C1956346   \n",
       "369551     rs9999653          4   55632082  1.000  0.08  C2362324   \n",
       "369552     rs9999820          4  155197173  0.925  0.08  C0025322   \n",
       "369553     rs9999820          4  155197173  0.925  0.08  C0085215   \n",
       "\n",
       "                                              diseaseName diseaseType  \\\n",
       "0                              Eosinophil count procedure   phenotype   \n",
       "1                             Leukemia, Myelocytic, Acute     disease   \n",
       "2       SPINAL MUSCULAR ATROPHY WITH RESPIRATORY DISTR...     disease   \n",
       "3                                         Body mass index   phenotype   \n",
       "4                                        Arthritis, Gouty     disease   \n",
       "...                                                   ...         ...   \n",
       "369549                        Acute myocardial infarction     disease   \n",
       "369550                            Coronary Artery Disease     disease   \n",
       "369551                                  Pediatric Obesity     disease   \n",
       "369552                                Premature Menopause     disease   \n",
       "369553                         Ovarian Failure, Premature     disease   \n",
       "\n",
       "       diseaseClass   diseaseSemanticType  score   EI  YearInitial  YearFinal  \\\n",
       "0               NaN  Laboratory Procedure   0.70  1.0       2016.0     2016.0   \n",
       "1               C04    Neoplastic Process   0.70  1.0       2017.0     2017.0   \n",
       "2       C16;C08;C10   Disease or Syndrome   0.70  1.0       2007.0     2007.0   \n",
       "3               NaN    Clinical Attribute   0.70  1.0       2019.0     2019.0   \n",
       "4       C16;C18;C05   Disease or Syndrome   0.70  1.0       2010.0     2013.0   \n",
       "...             ...                   ...    ...  ...          ...        ...   \n",
       "369549      C23;C14   Disease or Syndrome   0.01  0.0       2012.0     2012.0   \n",
       "369550          C14   Disease or Syndrome   0.01  0.0       2007.0     2007.0   \n",
       "369551      C23;C18   Disease or Syndrome   0.01  1.0       2017.0     2017.0   \n",
       "369552          C13   Disease or Syndrome   0.01  1.0       2018.0     2018.0   \n",
       "369553      C13;C19   Disease or Syndrome   0.70  1.0       2018.0     2018.0   \n",
       "\n",
       "        NofPmids            source  \n",
       "0              1  GWASCAT;INFERRED  \n",
       "1              1  GWASCAT;INFERRED  \n",
       "2              1  CLINVAR;INFERRED  \n",
       "3              1  GWASCAT;INFERRED  \n",
       "4              3   GWASDB;INFERRED  \n",
       "...          ...               ...  \n",
       "369549         1            BEFREE  \n",
       "369550         1            BEFREE  \n",
       "369551         1            BEFREE  \n",
       "369552         1            BEFREE  \n",
       "369553         1  GWASCAT;INFERRED  \n",
       "\n",
       "[369554 rows x 16 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "59\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['Advanced breast cancer',\n",
       "       'Anatomic Stage II Breast Cancer AJCC v8',\n",
       "       'Anatomic Stage III Breast Cancer AJCC v8',\n",
       "       'Basal-Like Breast Carcinoma', 'Breast Cancer, Familial',\n",
       "       'Breast Carcinoma', 'Breast adenocarcinoma',\n",
       "       'Breast cancer stage II', 'Breast cancer stage III',\n",
       "       'Breast cancer, lobular', 'Carcinoma breast stage IV',\n",
       "       'Carcinoma of Male Breast', 'Contralateral breast cancer',\n",
       "       'Ductal Breast Carcinoma', 'Early-Stage Breast Carcinoma',\n",
       "       'Female Breast Carcinoma', 'HER-2 positive breast cancer',\n",
       "       'HER2-negative breast cancer', 'HER2-positive carcinoma of breast',\n",
       "       'Hereditary Breast and Ovarian Cancer Syndrome',\n",
       "       'Infiltrating duct carcinoma of female breast',\n",
       "       'Intermediate Grade Ductal Breast Carcinoma In Situ',\n",
       "       'Invasive Ductal Breast Carcinoma', 'Invasive carcinoma of breast',\n",
       "       'Lobular carcinoma in situ of breast',\n",
       "       'Locally advanced breast cancer', 'Luminal A Breast Carcinoma',\n",
       "       'Luminal B Breast Carcinoma', 'Malignant melanoma of breast',\n",
       "       'Malignant neoplasm of breast',\n",
       "       'Malignant neoplasm of female breast',\n",
       "       'Malignant neoplasm of male breast',\n",
       "       'Malignant phyllodes tumor of breast',\n",
       "       'Medullary carcinoma of breast', 'Metastatic Breast Carcinoma',\n",
       "       'Multifocal breast carcinoma', 'Node-negative breast cancer',\n",
       "       'Node-positive breast cancer',\n",
       "       'Oestrogen receptor positive breast cancer',\n",
       "       'Premenopausal breast cancer',\n",
       "       'Prognostic Stage II Breast Cancer AJCC v8',\n",
       "       'Prognostic Stage III Breast Cancer AJCC v8',\n",
       "       'Secondary malignant neoplasm of female breast',\n",
       "       'Sporadic Breast Carcinoma',\n",
       "       'Stage 0 Breast Cancer AJCC v6 and v7', 'Stage 0 Breast Carcinoma',\n",
       "       'Stage II Breast Cancer AJCC v6 and v7',\n",
       "       'Stage III Breast Cancer AJCC v6',\n",
       "       'Stage III Breast Cancer AJCC v7',\n",
       "       'Triple Negative Breast Neoplasms',\n",
       "       'Triple-Negative Breast Cancer Finding',\n",
       "       'Triple-Negative Breast Carcinoma', 'Unilateral Breast Carcinoma',\n",
       "       'Unilateral Breast Neoplasms', 'bilateral breast cancer',\n",
       "       'estrogen receptor-negative breast cancer',\n",
       "       'malignant neoplasm of breast staging',\n",
       "       'progesterone receptor-negative breast cancer',\n",
       "       'progesterone receptor-positive breast cancer'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Breast = list()\n",
    "for i in range(len(variant_disease['diseaseName'])):\n",
    "    if 'Breast' in variant_disease['diseaseName'][i] and 'Atypical Ductal Breast Hyperplasia' not in variant_disease['diseaseName'][i] \\\n",
    "        and 'Breast Diffuse Large B-Cell Lymphoma' not in variant_disease['diseaseName'][i] \\\n",
    "        and 'Breast Diseases' not in variant_disease['diseaseName'][i] \\\n",
    "        and 'Breast Fibrocystic Disease' not in variant_disease['diseaseName'][i] \\\n",
    "        and 'Breast size' not in variant_disease['diseaseName'][i] \\\n",
    "        and 'Columnar Cell Change of the Breast' not in variant_disease['diseaseName'][i] \\\n",
    "        and 'Columnar Cell Hyperplasia of the Breast' not in variant_disease['diseaseName'][i] \\\n",
    "        and 'Mammographic Breast Density' not in variant_disease['diseaseName'][i] \\\n",
    "        and 'Zinc Deficiency, Neonatal, due to Low Breast Milk Zinc' not in variant_disease['diseaseName'][i]:\n",
    "        Breast.append(i)\n",
    "    if 'breast' in variant_disease['diseaseName'][i] :\n",
    "        Breast.append(i)\n",
    "\n",
    "print(len(np.unique(variant_disease['diseaseName'][Breast])))\n",
    "np.unique(variant_disease['diseaseName'][Breast])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snpId</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>DSI</th>\n",
       "      <th>DPI</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>diseaseName</th>\n",
       "      <th>diseaseType</th>\n",
       "      <th>diseaseClass</th>\n",
       "      <th>diseaseSemanticType</th>\n",
       "      <th>score</th>\n",
       "      <th>EI</th>\n",
       "      <th>YearInitial</th>\n",
       "      <th>YearFinal</th>\n",
       "      <th>NofPmids</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>rs1001179</td>\n",
       "      <td>11</td>\n",
       "      <td>34438684</td>\n",
       "      <td>0.641</td>\n",
       "      <td>0.68</td>\n",
       "      <td>C0006142</td>\n",
       "      <td>Malignant neoplasm of breast</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C17</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.02</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>2</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>rs1001179</td>\n",
       "      <td>11</td>\n",
       "      <td>34438684</td>\n",
       "      <td>0.641</td>\n",
       "      <td>0.68</td>\n",
       "      <td>C0678222</td>\n",
       "      <td>Breast Carcinoma</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C17</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.02</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>2</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>rs10012</td>\n",
       "      <td>2</td>\n",
       "      <td>38075247</td>\n",
       "      <td>0.716</td>\n",
       "      <td>0.28</td>\n",
       "      <td>C0006142</td>\n",
       "      <td>Malignant neoplasm of breast</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C17</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>3</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>rs10012</td>\n",
       "      <td>2</td>\n",
       "      <td>38075247</td>\n",
       "      <td>0.716</td>\n",
       "      <td>0.28</td>\n",
       "      <td>C0678222</td>\n",
       "      <td>Breast Carcinoma</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C17</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>3</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>rs10022462</td>\n",
       "      <td>4</td>\n",
       "      <td>88322666</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.08</td>\n",
       "      <td>C0678222</td>\n",
       "      <td>Breast Carcinoma</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C17</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>1</td>\n",
       "      <td>GWASCAT;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369507</th>\n",
       "      <td>rs999737</td>\n",
       "      <td>14</td>\n",
       "      <td>68567965</td>\n",
       "      <td>0.776</td>\n",
       "      <td>0.20</td>\n",
       "      <td>C0235653</td>\n",
       "      <td>Malignant neoplasm of female breast</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C17</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369508</th>\n",
       "      <td>rs999737</td>\n",
       "      <td>14</td>\n",
       "      <td>68567965</td>\n",
       "      <td>0.776</td>\n",
       "      <td>0.20</td>\n",
       "      <td>C0242787</td>\n",
       "      <td>Malignant neoplasm of male breast</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C17;C19</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369509</th>\n",
       "      <td>rs999737</td>\n",
       "      <td>14</td>\n",
       "      <td>68567965</td>\n",
       "      <td>0.776</td>\n",
       "      <td>0.20</td>\n",
       "      <td>C0678222</td>\n",
       "      <td>Breast Carcinoma</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C17</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>11</td>\n",
       "      <td>BEFREE;GWASCAT;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369511</th>\n",
       "      <td>rs999737</td>\n",
       "      <td>14</td>\n",
       "      <td>68567965</td>\n",
       "      <td>0.776</td>\n",
       "      <td>0.20</td>\n",
       "      <td>C3539878</td>\n",
       "      <td>Triple Negative Breast Neoplasms</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C17</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369512</th>\n",
       "      <td>rs999737</td>\n",
       "      <td>14</td>\n",
       "      <td>68567965</td>\n",
       "      <td>0.776</td>\n",
       "      <td>0.20</td>\n",
       "      <td>C4722518</td>\n",
       "      <td>Triple-Negative Breast Carcinoma</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C17</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9149 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             snpId chromosome  position    DSI   DPI diseaseId  \\\n",
       "152      rs1001179         11  34438684  0.641  0.68  C0006142   \n",
       "176      rs1001179         11  34438684  0.641  0.68  C0678222   \n",
       "191        rs10012          2  38075247  0.716  0.28  C0006142   \n",
       "201        rs10012          2  38075247  0.716  0.28  C0678222   \n",
       "334     rs10022462          4  88322666  1.000  0.08  C0678222   \n",
       "...            ...        ...       ...    ...   ...       ...   \n",
       "369507    rs999737         14  68567965  0.776  0.20  C0235653   \n",
       "369508    rs999737         14  68567965  0.776  0.20  C0242787   \n",
       "369509    rs999737         14  68567965  0.776  0.20  C0678222   \n",
       "369511    rs999737         14  68567965  0.776  0.20  C3539878   \n",
       "369512    rs999737         14  68567965  0.776  0.20  C4722518   \n",
       "\n",
       "                                diseaseName diseaseType diseaseClass  \\\n",
       "152            Malignant neoplasm of breast     disease      C04;C17   \n",
       "176                        Breast Carcinoma     disease      C04;C17   \n",
       "191            Malignant neoplasm of breast     disease      C04;C17   \n",
       "201                        Breast Carcinoma     disease      C04;C17   \n",
       "334                        Breast Carcinoma     disease      C04;C17   \n",
       "...                                     ...         ...          ...   \n",
       "369507  Malignant neoplasm of female breast     disease      C04;C17   \n",
       "369508    Malignant neoplasm of male breast     disease  C04;C17;C19   \n",
       "369509                     Breast Carcinoma     disease      C04;C17   \n",
       "369511     Triple Negative Breast Neoplasms     disease      C04;C17   \n",
       "369512     Triple-Negative Breast Carcinoma     disease      C04;C17   \n",
       "\n",
       "       diseaseSemanticType  score        EI  YearInitial  YearFinal  NofPmids  \\\n",
       "152     Neoplastic Process   0.02  1.000000       2012.0     2016.0         2   \n",
       "176     Neoplastic Process   0.02  1.000000       2012.0     2016.0         2   \n",
       "191     Neoplastic Process   0.03  0.666667       2009.0     2015.0         3   \n",
       "201     Neoplastic Process   0.03  0.666667       2009.0     2015.0         3   \n",
       "334     Neoplastic Process   0.70  1.000000       2017.0     2017.0         1   \n",
       "...                    ...    ...       ...          ...        ...       ...   \n",
       "369507  Neoplastic Process   0.01  1.000000       2016.0     2016.0         1   \n",
       "369508  Neoplastic Process   0.01  1.000000       2016.0     2016.0         1   \n",
       "369509  Neoplastic Process   0.77  0.909091       2009.0     2017.0        11   \n",
       "369511  Neoplastic Process   0.01  1.000000       2011.0     2011.0         1   \n",
       "369512  Neoplastic Process   0.01  1.000000       2011.0     2011.0         1   \n",
       "\n",
       "                         source  \n",
       "152                      BEFREE  \n",
       "176                      BEFREE  \n",
       "191                      BEFREE  \n",
       "201                      BEFREE  \n",
       "334            GWASCAT;INFERRED  \n",
       "...                         ...  \n",
       "369507                   BEFREE  \n",
       "369508                   BEFREE  \n",
       "369509  BEFREE;GWASCAT;INFERRED  \n",
       "369511                   BEFREE  \n",
       "369512                   BEFREE  \n",
       "\n",
       "[9149 rows x 16 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#These are information about the diseases from the database\n",
    "variant_disease.loc[Breast]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1697\n"
     ]
    }
   ],
   "source": [
    "#Map the database to the geneSymbol names\n",
    "snpId = variant_disease.loc[Breast]['snpId']\n",
    "variant_gene = pd.read_csv(\"variant_to_gene_mappings.tsv.gz\", compression='gzip', sep = '\\t')\n",
    "chosen = list()\n",
    "for i in range(len(variant_gene['snpId'])):\n",
    "    if variant_gene['snpId'][i] in np.array(snpId):\n",
    "        chosen.append(i)\n",
    "variant_gene.loc[chosen]\n",
    "\n",
    "# Here is the list of GWAS genes that we are interested in\n",
    "GWAS_Breast = np.unique(variant_gene.loc[chosen]['geneSymbol'])\n",
    "print(len(GWAS_Breast))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''GWAS enrichment'''\n",
    "'''DEGs are selected DSs'''\n",
    "    \n",
    "#DEGs of each celltype \n",
    "cancer_list = ['Lung','Liver','Breast','Colon','Ovary']\n",
    "cancer = \"Breast\" # change here for different cancers\n",
    "\n",
    "with open('/Users/cynthia_ye/Documents/OneDrive - Linköpings universitet/cancer/datafile/outputs_5cancers/new_outputs/GWAS/2022Mar09/GWAS_enrich_percelltype_'+cancer+'.txt','w') as f:\n",
    "\n",
    "    DEGs = pd.read_csv('/Users/cynthia_ye/Documents/OneDrive - Linköpings universitet/cancer/datafile/outputs_5cancers/new_outputs/MCDM/Ligand_target_allDS_of_allUR/ligand_target_merged_withFC.csv',sep = ',')\n",
    "    DEGs = DEGs.loc[DEGs['cancer']==cancer]\n",
    "    DEGs\n",
    "\n",
    "    celltype_DEG = np.unique(DEGs['receiver'])\n",
    "    print(\"celltypes_DEGs: \", celltype_DEG, file =f)\n",
    "\n",
    "    '''all_genes means the combine of all background genes in all celltypes'''\n",
    "    bg_file_path = (\"/Users/cynthia_ye/Documents/OneDrive - Linköpings universitet/cancer/datafile/outputs_5cancers/new_outputs/DEGs/Background_genes/background_genes_\"+cancer+\".csv\")\n",
    "    all_gene = np.arange(0)\n",
    "    all = pd.read_csv(bg_file_path,sep = ',') \n",
    "    all = all.drop('Unnamed: 0',axis = 1)\n",
    "\n",
    "    celltype_bg = all.columns\n",
    "    print(\"celltypes_bgs: \", celltype_bg, file =f)\n",
    "    DEGs\n",
    "\n",
    "    '''for Breast only - loop all celltype'''\n",
    "    for c in range(len(celltype_DEG)):\n",
    "        print('\\n', file =f)\n",
    "        print(celltype_DEG[c], file =f)\n",
    "        iDEGs = DEGs.loc[DEGs['receiver']==celltype_DEG[c]]['target'].unique()\n",
    "        print(celltype_DEG[c],\"DEGs\",len(iDEGs), file =f)\n",
    "        ibackground_genes = all[celltype_DEG[c]].unique()\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'NAN']\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'nan']\n",
    "        print(celltype_DEG[c],\"background\",len(ibackground_genes), file =f)\n",
    "\n",
    "        GWAS = list()\n",
    "        for g in GWAS_Breast:\n",
    "            for a in ibackground_genes:\n",
    "                a = str(a)\n",
    "                for i in range(len(a.split('///'))):\n",
    "                    if g == a.split('///')[i]:\n",
    "                        GWAS.append(a)                   \n",
    "        GWAS = np.unique(GWAS)\n",
    "        GWAS\n",
    "\n",
    "        nonDEGs = list(set(ibackground_genes) - set(iDEGs))                \n",
    "        table11 = len(np.intersect1d(GWAS, iDEGs))\n",
    "        table12 = len(iDEGs) - table11\n",
    "        table21 = len(np.intersect1d(GWAS, nonDEGs))\n",
    "        table22 = len(ibackground_genes) - (table21 + table12 + table11)\n",
    "        table = pd.DataFrame([[table11, table12], [table21, table22]])\n",
    "        table.index = ('DEGs', 'nonDEGs')\n",
    "        table.columns = ('GWAS', 'nonGWAS')\n",
    "\n",
    "        print('Breast Cancer - ',celltype_DEG[c], file =f )\n",
    "        print(table, file =f)\n",
    "        odds, pval = stats.fisher_exact(table)\n",
    "        print(\"Odds ratio: \", round(odds, 4), file =f)\n",
    "        print(\"p-value: \",round(pval, 6), file =f)\n",
    "\n",
    "        # if the odds are > 1 and pval <0.05 The list of DEGs is significantly enriched by the GWAS genes\n",
    "        # if the odds are < 1 and pval <0.05 The list of DEGs is not enriched by the GWAS genes\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''GWAS enrichment'''\n",
    "'''DEGs are selected all DSs\n",
    "    background genes are all differentially expressed genes\n",
    "'''\n",
    "    \n",
    "#DEGs of each celltype \n",
    "cancer_list = ['Lung','Liver','Breast','Colon','Ovary']\n",
    "cancer = \"Breast\" # change here for different cancers\n",
    "\n",
    "with open('/Users/cynthia_ye/Documents/OneDrive - Linköpings universitet/cancer/datafile/outputs_5cancers/new_outputs/GWAS/2022Mar09/GWAS_enrich_percelltype_DEGasBG_'+cancer+'.txt','w') as f:\n",
    "\n",
    "    DEGs = pd.read_csv('/Users/cynthia_ye/Documents/OneDrive - Linköpings universitet/cancer/datafile/outputs_5cancers/new_outputs/MCDM/Ligand_target_allDS_of_allUR/ligand_target_merged_withFC.csv',sep = ',')\n",
    "    DEGs = DEGs.loc[DEGs['cancer']==cancer]\n",
    "    DEGs\n",
    "\n",
    "    celltype_DEG = np.unique(DEGs['receiver'])\n",
    "    print(\"celltypes_DEGs: \", celltype_DEG, file =f)\n",
    "\n",
    "    '''Differentially expressed genes in each cell type were background genes'''\n",
    "    bg_file_path = (\"/Users/cynthia_ye/Documents/OneDrive - Linköpings universitet/cancer/datafile/outputs_5cancers/new_outputs/DEGs/DEGs_of_\"+cancer+\"/DEGs_merged_\"+cancer+\"_MAST.txt\")\n",
    "    all_gene = np.arange(0)\n",
    "    all = pd.read_csv(bg_file_path,sep = '\\t') \n",
    "    all\n",
    "    \n",
    "    celltype_bg = all.columns\n",
    "    print(\"celltypes_bgs: \", celltype_bg, file =f)\n",
    "    DEGs\n",
    "\n",
    "    '''loop all celltype'''\n",
    "    for c in range(len(celltype_DEG)):\n",
    "        print('\\n', file =f)\n",
    "        print(celltype_DEG[c], file =f)\n",
    "        iDEGs = DEGs.loc[DEGs['receiver']==celltype_DEG[c]]['target'].unique()\n",
    "        print(celltype_DEG[c],\"DEGs\",len(iDEGs), file =f)\n",
    "        ibackground_genes = all[celltype_DEG[c]].unique()\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'NAN']\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'nan']\n",
    "        print(celltype_DEG[c],\"background\",len(ibackground_genes), file =f)\n",
    "\n",
    "        GWAS = list()\n",
    "        for g in GWAS_Breast:\n",
    "            for a in ibackground_genes:\n",
    "                a = str(a)\n",
    "                for i in range(len(a.split('///'))):\n",
    "                    if g == a.split('///')[i]:\n",
    "                        GWAS.append(a)                   \n",
    "        GWAS = np.unique(GWAS)\n",
    "        GWAS\n",
    "\n",
    "        nonDEGs = list(set(ibackground_genes) - set(iDEGs))                \n",
    "        table11 = len(np.intersect1d(GWAS, iDEGs))\n",
    "        table12 = len(iDEGs) - table11\n",
    "        table21 = len(np.intersect1d(GWAS, nonDEGs))\n",
    "        table22 = len(ibackground_genes) - (table21 + table12 + table11)\n",
    "        table = pd.DataFrame([[table11, table12], [table21, table22]])\n",
    "        table.index = ('DEGs', 'nonDEGs')\n",
    "        table.columns = ('GWAS', 'nonGWAS')\n",
    "\n",
    "        print('Breast Cancer - ',celltype_DEG[c], file =f )\n",
    "        print(table, file =f)\n",
    "        odds, pval = stats.fisher_exact(table)\n",
    "        print(\"Odds ratio: \", round(odds, 4), file =f)\n",
    "        print(\"p-value: \",round(pval, 6), file =f)\n",
    "\n",
    "        # if the odds are > 1 and pval <0.05 The list of DEGs is significantly enriched by the GWAS genes\n",
    "        # if the odds are < 1 and pval <0.05 The list of DEGs is not enriched by the GWAS genes\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ALDH1A1', 'BAX', 'CCND1', 'COL18A1', 'COL1A1', 'FN1', 'GNAS',\n",
       "       'KRT8', 'MMP2', 'MMP9', 'MUC1', 'NFKBIA', 'NR3C1', 'RUNX1', 'SHC1',\n",
       "       'SPP1', 'TGFB1', 'ZFP36'], dtype=object)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list GWAS gene that in our gene-of-interests\n",
    "np.intersect1d(GWAS, DEGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''GWAS enrichment'''\n",
    "'''DEGs are selected all DSs of the top scURs\n",
    "    background genes are all differentially expressed genes\n",
    "'''\n",
    "    \n",
    "#DEGs of each celltype \n",
    "cancer_list = ['Lung','Liver','Breast','Colon','Ovary']\n",
    "cancer = \"Breast\" # change here for different cancers\n",
    "\n",
    "with open('/Users/cynthia_ye/Documents/OneDrive - Linköpings universitet/cancer/datafile/outputs_5cancers/new_outputs/GWAS/2022Mar09/GWAS_enrich_percelltype_DS4topscUR_DEGasBG_'+cancer+'.txt','w') as f:\n",
    "\n",
    "    DEGs = pd.read_csv('/Users/cynthia_ye/Documents/OneDrive - Linköpings universitet/cancer/datafile/outputs_5cancers/new_outputs/MCDM/ligand_target_merged_DS_of_scUR.csv',sep = ',')\n",
    "    top_scURs = [\"COL1A1\",\"FN1\",\"SPP1\",\"COL4A1\",\"COL18A1\", \"PLAU\" , \"CLEC11A\" ,\"MDK\"]\n",
    "    DEGs = DEGs.loc[np.isin(DEGs['ligand'], top_scURs)]\n",
    "    DEGs = DEGs.loc[DEGs['cancer']==cancer]\n",
    "\n",
    "    celltype_DEG = np.unique(DEGs['receiver'])\n",
    "    print(\"celltypes_DEGs: \", celltype_DEG, file =f)\n",
    "\n",
    "    '''Differentially expressed genes in each cell type were background genes'''\n",
    "    bg_file_path = (\"/Users/cynthia_ye/Documents/OneDrive - Linköpings universitet/cancer/datafile/outputs_5cancers/new_outputs/DEGs/DEGs_of_\"+cancer+\"/DEGs_merged_\"+cancer+\"_MAST.txt\")\n",
    "    all_gene = np.arange(0)\n",
    "    all = pd.read_csv(bg_file_path,sep = '\\t') \n",
    "    all\n",
    "    \n",
    "    celltype_bg = all.columns\n",
    "    print(\"celltypes_bgs: \", celltype_bg, file =f)\n",
    "    DEGs\n",
    "\n",
    "    '''loop all celltype'''\n",
    "    for c in range(len(celltype_DEG)):\n",
    "        print('\\n', file =f)\n",
    "        print(celltype_DEG[c], file =f)\n",
    "        iDEGs = DEGs.loc[DEGs['receiver']==celltype_DEG[c]]['target'].unique()\n",
    "        print(celltype_DEG[c],\"DEGs\",len(iDEGs), file =f)\n",
    "        ibackground_genes = all[celltype_DEG[c]].unique()\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'NAN']\n",
    "        ibackground_genes = [x for x in ibackground_genes if x != 'nan']\n",
    "        print(celltype_DEG[c],\"background\",len(ibackground_genes), file =f)\n",
    "\n",
    "        GWAS = list()\n",
    "        for g in GWAS_Breast:\n",
    "            for a in ibackground_genes:\n",
    "                a = str(a)\n",
    "                for i in range(len(a.split('///'))):\n",
    "                    if g == a.split('///')[i]:\n",
    "                        GWAS.append(a)                   \n",
    "        GWAS = np.unique(GWAS)\n",
    "        GWAS\n",
    "\n",
    "        nonDEGs = list(set(ibackground_genes) - set(iDEGs))                \n",
    "        table11 = len(np.intersect1d(GWAS, iDEGs))\n",
    "        table12 = len(iDEGs) - table11\n",
    "        table21 = len(np.intersect1d(GWAS, nonDEGs))\n",
    "        table22 = len(ibackground_genes) - (table21 + table12 + table11)\n",
    "        table = pd.DataFrame([[table11, table12], [table21, table22]])\n",
    "        table.index = ('DEGs', 'nonDEGs')\n",
    "        table.columns = ('GWAS', 'nonGWAS')\n",
    "\n",
    "        print('Breast Cancer - ',celltype_DEG[c], file =f )\n",
    "        print(table, file =f)\n",
    "        odds, pval = stats.fisher_exact(table)\n",
    "        print(\"Odds ratio: \", round(odds, 4), file =f)\n",
    "        print(\"p-value: \",round(pval, 6), file =f)\n",
    "\n",
    "        # if the odds are > 1 and pval <0.05 The list of DEGs is significantly enriched by the GWAS genes\n",
    "        # if the odds are < 1 and pval <0.05 The list of DEGs is not enriched by the GWAS genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
